crates/bevy_ecs/src/storage/blob_array.rs

```diff
@@ -0,0 +1,455 @@
+            self.data = NonNull::new(new_data).unwrap_or_else(|| handle_alloc_error(new_layout));
+        }
+    }
+
+    /// Reallocate memory for this array.
+    /// For example, if the length (number of stored elements) reached the capacity (number of elements the current allocation can store),
+    /// you might want to use this method to increase the allocation, so more data can be stored in the array.
+    ///
+    /// # Safety
+    /// - `current_capacity` + `increment` doesn't overflow `usize`
+    /// - The size of the resulting array does not overflow `usize` (specifically, see the safety requirements of [`array_layout_unchecked`])
+    /// - `current_capacity` is indeed the current capacity of this array.
+    /// - After calling this method, the caller must update their saved capacity to reflect the change.
+    pub(super) unsafe fn realloc(
+        &mut self,
+        current_capacity: NonZeroUsize,
+        new_capacity: NonZeroUsize,
+    ) {
+        if !self.is_zst() {
+            let new_layout =
+            // SAFETY: Safety requirement 2
+            unsafe {
+                array_layout_unchecked(&self.item_layout, new_capacity.get())
+            };
+            // SAFETY:
+            // - ptr was be allocated via this allocator
+            // - the layout used to previously allocate this array is equivalent to `array_layout(&self.item_layout, current_capacity.get())`
+            // - `item_layout.size() > 0` (`self.is_zst`==false) and `new_capacity > 0` (incrememt>0), so the layout size is non-zero
+            // - "new_size, when rounded up to the nearest multiple of layout.align(), must not overflow (i.e., the rounded value must be less than usize::MAX)",
+            // since the item size is always a multiple of its align, the rounding cannot happen
+            // here and the overflow is handled in `array_layout`
+            let new_data = std::alloc::realloc(
+                self.get_ptr_mut().as_ptr(),
+                // SAFETY: This is the Layout of the current array, it must be valid, if it hadn't have been, there would have been a panic on a previous allocation
+                array_layout_unchecked(&self.item_layout, current_capacity.get()),
+                new_layout.size(),
+            );
+            self.data = NonNull::new(new_data).unwrap_or_else(|| handle_alloc_error(new_layout));
+        }
+    }
+
+    /// Initializes the value at `index` to `value`. This function does not do any bounds checking.
+    ///
+    /// # Safety
+    /// - `index` must be in bounds (`index` < `len`)
+    /// - the memory in the [`BlobArray`] starting at index `index`, of a size matching this [`BlobArray`]'s `item_layout`, must have been previously allocated.
+    #[inline]
+    pub unsafe fn initialize_unchecked(&mut self, index: usize, value: OwningPtr<'_>) {
+        let ptr = self.get_unchecked_mut(index);
+        std::ptr::copy_nonoverlapping::<u8>(value.as_ptr(), ptr.as_ptr(), self.item_layout.size());
```

Comment:
This feels like an operation that should be on `OwningPtr`/`PtrMut`

Reply:
Do you mean implement a custom `copy_nonoverlapping` that takes in `OwningPtr` / `PtrMut`?

Reply:
Yes, I imagine it would take a `OwningPtr`, since it takes ownership of the value being written, and a `PtrMut`, since it writes through the other pointer.